libs/domain/feature-loaders/src/snapshot/SnapshotProposalLoader.ts

```diff
@@ -0,0 +1,187 @@
+    `;
+}
+
+// Helper for optional fields
+const optional = <T extends t.Mixed>(x:T)=>t.union([x,t.undefined])
+
+export const ProposalCodec = t.strict({
+    id: t.string,
+    author: t.string,
+    created: t.number,
+    space: optional(t.strict({
+        id: t.string,
+        name: t.string
+    })),
+    type: optional(t.string),
+    strategies: withMessage(t.array(t.strict({
+        name:t.string,
+    })),()=>"strategies failed"),
+    title: t.string,
+    body: optional(t.string),
+    choices: t.array(t.string),
+    start: t.number,
+    end: t.number,
+    snapshot: t.string,
+    state: t.string,
+    link: optional(t.string),
+    scores: optional(t.array(t.number)),
+    votes: optional(t.number)
+});
+
+export const ProposalsCodec = t.strict({
+    proposals: t.array(ProposalCodec),
+});
+
+export type Proposals = t.TypeOf<typeof ProposalsCodec>;
+
+const INFO = {
+    namespace: "snapshot",
+    name: "Proposal",
+    version: "V1",
+};
+@Nested()
+class Space {
+    @OptionalProperty()
+    name: string
+    @NonEmptyProperty()
+    id: string
+}
+
+@Nested()
+class Strategy {
+    @NonEmptyProperty()
+    name:string
+}
+
+@Data({
+    info: INFO,
+})
+export class Proposal {
+    @NonEmptyProperty()
+    id: string
+    @NonEmptyProperty()
+    author: string
+    @NonEmptyProperty()
+    created: number
+    @OptionalObjectRef(Space)
+    space: Space | undefined
+    @OptionalProperty()
+    type:string | undefined
+    @RequiredArrayRef(Strategy)
+    strategies: Strategy[]
+    @NonEmptyProperty()
+    title:string
+    @OptionalProperty()
+    body: string | undefined
+    @RequiredStringArrayOf()
+    choices: string[]
+    @NonEmptyProperty()
+    start: number
+    @NonEmptyProperty()
+    end: number
+    @NonEmptyProperty()
+    snapshot: string
+    @NonEmptyProperty()
+    state: string
+    @OptionalProperty()
+    link: string | undefined
+    @OptionalNumberArrayOf()
+    scores: number[] | undefined
+    @OptionalProperty()
+    votes: number 
+}
+
+export class SnapshotProposalLoader extends GraphQLDataLoaderBase<Proposals, Proposal> {
+    public info = INFO;
+
+    protected cursorMode = "cursor" as const;
+    protected cursor: number;
+
+    protected batchSize = BATCH_SIZE;
+    protected type = Proposal;
+    protected cadenceConfig = {
+        [ScheduleMode.BACKFILL]: { seconds: 5 },
+        [ScheduleMode.INCREMENTAL]: { minutes: 5 },
+    };
+
+    protected graphQLQuery: DocumentNode;
+    protected codec = ProposalsCodec;
+
+    constructor(client: GraphQLClient,space: String) {
+        super(client);
+        this.graphQLQuery = makeQUERY(space)
+    }
+
+    protected preLoad(
+        context: InternalLoadContext
+    ): TE.TaskEither<ProgramError, InternalLoadContext> {
+        // Ensure a number cursor is used
+        if (!context.cursor) {
+            context.cursor = 0
+        }
+        if (typeof context.cursor == "string") {
+            context.cursor = Number(context.cursor)
```

Comment:
We're returning a `TaskEither` from this function which means that it is an asynchronous operation that never fails (doesn't throw an exception). In this case it is possible that `context.cursor` can't be parsed as a `Number` and an exception will be thrown. In order to prevent this fp-ts gives us `tryCatch` that works like this:

```ts
TE.tryCatch(
    async () => Number(context.cursor),
    (e: unknown) => {
        return new UnknownError(e);
    }
);
```
Can you wrap this code above with a `tryCatch`? 👆

Reply:
This whole function was removed